src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -0,0 +1,449 @@
+			exit (FATAL_EXIT_NO_ASSEMBLIES);
+		}
+
+		if (!zip_adjust_data_offset (fd, local_header_offset, data_offset)) {
+			log_fatal (LOG_DEFAULT, "Failed to adjust data start offset for entry %u in APK file %s", i, apk_name);
+			exit (FATAL_EXIT_NO_ASSEMBLIES);
+		}
+#ifdef DEBUG
+		log_warn (LOG_DEFAULT, "    ZIP: local header offset: %u; data offset: %u; file size: %u", local_header_offset, data_offset, file_size);
+#endif
+		if (compression_method != 0)
+			continue;
+
+#if defined (DEBUG)
+		if (utils.ends_with (file_name, ".jm")) {
+			md_mmap_info map_info   = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+			add_type_mapping (&java_to_managed_maps, apk_name, file_name, (const char*)map_info.area);
+			continue;
+		}
+		if (utils.ends_with (file_name, ".mj")) {
+			md_mmap_info map_info   = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+			add_type_mapping (&managed_to_java_maps, apk_name, file_name, (const char*)map_info.area);
+			continue;
+		}
+#endif
+		if (strncmp (prefix, file_name, prefix_len) != 0)
+			continue;
+
+		// assemblies must be 4-byte aligned, or Bad Things happen
+		if ((data_offset & 0x3) != 0) {
+			log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", file_name, data_offset);
+			log_fatal (LOG_ASSEMBLY, "You MUST run `zipalign` on %s\n", strrchr (apk_name, '/') + 1);
+			exit (FATAL_EXIT_MISSING_ZIPALIGN);
+		}
+
+		bool entry_is_overridden = !should_register (strrchr (file_name, '/') + 1);
+
+		if ((utils.ends_with (file_name, ".mdb") || utils.ends_with (file_name, ".pdb")) &&
+				register_debug_symbols &&
+				!entry_is_overridden &&
+				bundled_assemblies != nullptr) {
+			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, file_name, apk_name);
+			if (register_debug_symbols_for_assembly (file_name, (bundled_assemblies) [bundled_assemblies_count - 1],
+						(const mono_byte*)map_info.area,
+				        static_cast<int>(file_size)))
+				continue;
+		}
+
+		if (utils.ends_with (file_name, ".config") && bundled_assemblies != nullptr) {
+			char *assembly_name = strdup (basename (file_name));
+			// Remove '.config' suffix
+			*strrchr (assembly_name, '.') = '\0';
+
+			md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+			mono_register_config_for_assembly (assembly_name, (const char*)map_info.area);
+
+			continue;
+		}
+
+		if (!(utils.ends_with (file_name, ".dll") || utils.ends_with (file_name, ".exe")))
+			continue;
+
+		if (entry_is_overridden)
+			continue;
+
+		size_t alloc_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, sizeof(void*), bundled_assemblies_count + 1);
+		bundled_assemblies = reinterpret_cast<MonoBundledAssembly**> (utils.xrealloc (bundled_assemblies, alloc_size));
+		MonoBundledAssembly *cur = bundled_assemblies [bundled_assemblies_count] = reinterpret_cast<MonoBundledAssembly*> (utils.xcalloc (1, sizeof (MonoBundledAssembly)));
```

Comment:
Why `xcalloc()` and not `new`?

Reply:
1. because of `xrealloc` above (`realloc`-like in C++ is trickier - you have to take into account constructors, make sure the type is POD etc - I *DO* plan to modify this code but in a separate PR. I deliberately left this code as it is in `master` to avoid introducing, for some reason, the same debugging problem PR #3606 has)